reflect_abi_patch.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"maps"
+	"os"
+	"slices"
+	"strings"
+)
+
+func abiNamePatch(path string) (string, error) {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return "", err
+	}
+
+	find := `return unsafe.String(n.DataChecked(1+i, "non-empty string"), l)`
+	replace := `name := unsafe.String(n.DataChecked(1+i, "non-empty string"), l)
+	return _realName(name)`
+
+	str := strings.Replace(string(data), find, replace, 1)
+
+	realname := `
+//go:linkname _realName
+func _realName(name string) string
+`
+
+	return str + realname, nil
+}
+
+var patchMainFile = ""
+
+func reflectMainPrePatch(path string) ([]byte, error) {
+	if patchMainFile != "" {
+		// already patched another file in main
+		return nil, nil
+	}
+
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	patchMainFile = path
+
+	nameMap := "\nvar _nameMap = map[string]string{}"
+
+	return append(content, []byte(realNameCode+nameMap)...), nil
+}
+
+func reflectMainPostPatch(file []byte, lpkg *listedPackage) []byte {
+	obfMapName := hashWithPackage(lpkg, "_nameMap")
+	nameMap := fmt.Sprintf("%s = map[string]string{", obfMapName)
+
+	values := ""
+	obfToReal := realForReflected()
+
+	keys := slices.Collect(maps.Keys(obfToReal))
+	slices.Sort(keys)
+
+	for _, obf := range keys {
+		values += fmt.Sprintf(`"%s": "%s",`, obf, obfToReal[obf])
+	}
+
+	return bytes.Replace(file, []byte(nameMap), []byte(nameMap+values), 1)
+}
+
+// this may look unnessecary for simpler cases but we need support complex cases like:
+// name := "*struct { AQ45rr68K string; ipq5aQSIqN string; hNfiW5O5LVq struct { gPTbGR00hu string } }"
+const realNameCode = `
+//go:linkname _realName internal/abi._realName
+func _realName(name string) string {
+	for i := 0; i < len(name); {
+		remLen := len(name[i:])
+
+		found := false
+		for obfName, real := range _nameMap {
+			keyLen := len(obfName)
+			if keyLen > remLen {
+				continue
+			}
+			if name[i:i+keyLen] == obfName {
+				name = name[:i] + real + name[i+keyLen:]
+
+				found = true
+				i += len(real)
+				break
+			}
+		}
+
+		if !found {
+			i++
+		}
+	}
+
+	return name
+}`
+
+// map reflected obfuscated names to real names
```

Comment:
document that the entries in this map are not separated or namespaced by package, because we assume that they are unique enough so that we don't have to.